File: spring-boot-rest-api/src/main/java/br/com/uabrestingaseca/biblioteca/model/Livro.java
package br.com.uabrestingaseca.biblioteca.model;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.LazyCollection;
import org.hibernate.annotations.LazyCollectionOption;

import javax.persistence.*;
import javax.validation.constraints.Max;
import javax.validation.constraints.Min;
import javax.validation.constraints.NotBlank;
import java.io.Serializable;
import java.util.LinkedList;
import java.util.List;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
@Entity
@Table(name = "livro")
public class Livro implements Serializable {

    private static final long serialVersionUID = 1L;

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;

    @NotBlank(message = "Título dever ser informado")
    private String titulo;

    private String resumo;

    private Long isbn;

    private String cutter;

    @ManyToOne
    private Editora editora;

    private String edicao;

    private String volume;

    @Column(name = "num_paginas")
    private Integer numPaginas;

    @ManyToOne
    private Assunto assunto;

    @Min(value = 0, message = "Ano inválido")
    @Max(value = 2100, message = "Ano inválido")
    @Column(name = "ano_publicacao")
    private Short anoPublicacao;

    @ManyToMany
    @JoinTable(name = "autor_livro",
            joinColumns = { @JoinColumn( name = "livro_id") },
            inverseJoinColumns = { @JoinColumn( name = "autor_id") }
    )
    @LazyCollection(LazyCollectionOption.FALSE)
    private List<Autor> autores;

    @ManyToMany
    @JoinTable(name = "categoria_livro",
            joinColumns = { @JoinColumn( name = "livro_id") },
            inverseJoinColumns = { @JoinColumn( name = "categoria_id") }
    )
    @LazyCollection(LazyCollectionOption.FALSE)
    private List<Categoria> categorias;

    @OneToMany
    @JoinColumn(name = "livro_id", insertable = false, updatable = false)
    @LazyCollection(LazyCollectionOption.FALSE)
    private List<Exemplar> exemplares;

    private String nomesAutores;

    public String getNomesAutores(){
        if (nomesAutores == null && autores != null){
            StringBuilder builder = new StringBuilder();
            autores.forEach(a->{
                if (builder.length() > 0){
                    builder.append("; ");
                }
                builder.append(a.getSobrenome().toUpperCase());
                builder.append(", ");
                builder.append(a.getNome());
            });
            nomesAutores = builder.toString();
        }
        return nomesAutores;
    }

    @Deprecated
    public Livro(Integer id){
        this.id = id;
    }

    public List<Autor> getAutores() {
        if (autores == null)
            autores = new LinkedList<>();
        return autores;
    }

    public void setAutores(List<Autor> autores) {
        this.autores = autores;
        getNomesAutores();
    }

    public List<Categoria> getCategorias() {
        if (categorias == null)
            categorias = new LinkedList<>();
        return categorias;
    }

    public void setCategorias(List<Categoria> categorias) {
        this.categorias = categorias;
    }

    public List<Exemplar> getExemplares() {
        if (exemplares == null)
            exemplares = new LinkedList<>();
        return exemplares;
    }

    public void setExemplares(List<Exemplar> exemplares) {
        this.exemplares = exemplares;
    }

}

File: biblioteca-react/src/model/Livro.ts
import { AxiosResponse } from 'axios';
import { Assunto }  from './Assunto';
import { Autor }  from './Autor';
import { Categoria }  from './Categoria';
import { Editora }  from './Editora';

export class Livro{

    constructor(
        public id?: number,
        public titulo?: string,
        public resumo?: string | null,
        public isbn?: number | null,
        public cutter?: string | null,
        public editora?: Editora,
        public edicao?: string | null,
        public volume?: string | null,
        public numPaginas?: number | null,
        public assunto?: Assunto,
        public anoPublicacao?: number | null,
        public autores?: Autor[],
        public categorias?: Categoria[]
    ){}

    private static _createLivro(data: any):Livro{
        const editora = new Editora();
        if (data.editora){
            editora.id = data.editora.id;
            editora.nome = data.editora.nome;
        }
        const assunto = new Assunto();
        if (data.assunto){
            assunto.id = data.assunto.id;
            assunto.descricao = data.assunto.descricao;
            assunto.cores = data.assunto.cores;
            assunto.cdu = data.assunto.cdu;
        }
        const autores = data.autores.map((a : any) =>{
            return new Autor(a.id, a.nome, a.sobrenome, a.info);
        });
        const categorias = data.categorias.map((c : any) => {
            return new Categoria(c.id, c.descricao);
        });
        return new Livro(
            data.id, data.titulo, data.resumo, data.isbn, data.cutter,
            editora, data.edicao, data.volume, data.numPaginas, assunto,
            data.anoPublicacao, autores, categorias
        );
    }

    public static createLivroFromState(state : any):Livro{
        if (!state)
            return new Livro();
        return this._createLivro(state);
    }

    public static createLivro(response: AxiosResponse<any>):Livro{
        const data = response.data;
        return this._createLivro(data);
    }

    public static createLivros(response: AxiosResponse<any>):Livro[]{
        const data = response.data;
        return data.map((d : any) => {
            return this._createLivro(d);
        });
    }

    public getNomesAutores():string{
        if (this.autores){
            return this.autores.map(a => {
                return `${a.sobrenome?.toUpperCase()}, ${a.nome}`          
            }).join('; ');
        }
        return '';
    }

    public getArrayNomesAutores(): Array<string>{
        if (this.autores){
            return this.autores.map(a => {
                return `${a.sobrenome?.toUpperCase()}, ${a.nome}`          
            });
        }
        return new Array<string>();
    }

}